### PR TITLE
Self-contained  Magic Link Authenticator

### PR DESCRIPTION
**Pull request by Yoshani**
### Purpose
In the current implementation, the MagicLink authenticator is dependent on the identifier first to authenticate and resolve the user. But ideally, we should improve the MagicLink authenticator to embed the functionality of the identifier first authenticator to resolve the user using a given identifier.

This PR introduces identifier first capabilities into the Magic Link Authenticator.

### Related Issues
- Issue https://github.com/wso2/product-is/issues/15145

### Related PRs
- PR https://github.com/wso2/carbon-identity-framework/pull/4337

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 1]
```diff
@@ -149,6 +165,184 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        Map<String, String> runtimeParams = getRuntimeParams(context);
+        String identifierFromRequest = request.getParameter(MagicLinkAuthenticatorConstants.USER_NAME);
+        if (StringUtils.isBlank(identifierFromRequest)) {
+            throw new InvalidCredentialsException(MagicLinkAuthErrorConstants.ErrorMessages.EMPTY_USERNAME.getCode(),
+                    MagicLinkAuthErrorConstants.ErrorMessages.EMPTY_USERNAME.getMessage());
+        }
+        if (runtimeParams != null) {
```
Comment (Yoshani) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:178]: map utils

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 2]
```diff
@@ -149,6 +165,184 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+                userId = resolvedUserResult.getUser().getUserID();
+            } else {
+                throw new InvalidCredentialsException(
+                        MagicLinkAuthErrorConstants.ErrorMessages.USER_DOES_NOT_EXISTS.getCode(),
+                        MagicLinkAuthErrorConstants.ErrorMessages.USER_DOES_NOT_EXISTS.getMessage(),
+                        org.wso2.carbon.identity.application.common.model.User.getUserFromUserName(username));
+            }
+        }
+
+        if (context.getCallerPath() != null && context.getCallerPath().startsWith("/t/")) {
```
Comment (Yoshani) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:216]: extract to pvt method

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 3]
```diff
@@ -149,6 +165,184 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+                                                MagicLinkAuthErrorConstants.ErrorMessages
+                                                        .USER_NOT_IDENTIFIED_IN_HIERARCHY.getCode()));
+                        tenantAwareUsername = user.getUsername();
+                        username = UserCoreUtil.addTenantDomainToEntry(tenantAwareUsername, user.getTenantDomain());
+                        userId = user.getUserID();
+                    }
+                } catch (OrganizationManagementException e) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("IdentifierHandler failed while trying to resolving user's resident org", e);
+                    }
```
Comment (Yoshani) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:240]: use diagnostic log

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 4]
```diff
@@ -149,6 +165,184 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+
+        String tenantDomain = MultitenantUtils.getTenantDomain(username);
+        Map<String, Object> authProperties = context.getProperties();
+        if (authProperties == null) {
+            authProperties = new HashMap<>();
+            context.setProperties(authProperties);
+        }
+
+        if (getAuthenticatorConfig().getParameterMap() != null) {
+            String validateUsername = getAuthenticatorConfig().getParameterMap().get("ValidateUsername");
```
Comment (Yoshani) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:266]: use constant

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 5]
```diff
@@ -149,6 +165,163 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+    private void processIdfAuthenticationResponse(HttpServletRequest request, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        Map<String, String> runtimeParams = getRuntimeParams(context);
+        String identifierFromRequest = request.getParameter(MagicLinkAuthenticatorConstants.USER_NAME);
+        if (StringUtils.isBlank(identifierFromRequest)) {
+            throw new InvalidCredentialsException(MagicLinkAuthErrorConstants.ErrorMessages.EMPTY_USERNAME.getCode(),
+                    MagicLinkAuthErrorConstants.ErrorMessages.EMPTY_USERNAME.getMessage());
+        }
+        if (MapUtils.isEmpty(runtimeParams)) {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:177]: shouldn't this be a "is not empty" check?

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 6]
```diff
@@ -149,6 +165,163 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+        String username = identifierFromRequest;
+        if (!IdentityUtil.isEmailUsernameValidationDisabled()) {
+            FrameworkUtils.validateUsername(identifierFromRequest, context);
+            username = FrameworkUtils.preprocessUsername(identifierFromRequest, context);
+        }
+
+        String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+        String userId = null;
+
+        // resolve user from multi attribute login
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:200]: Let's use sentence case for comments

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 7]
```diff
@@ -149,6 +165,163 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+            if (StringUtils.isNotBlank(requestTenantDomain) &&
+                    !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equalsIgnoreCase(requestTenantDomain)) {
+                User user = resolveUserFromOrganizationHierarchy(tenantAwareUsername, requestTenantDomain, username);
+                if (user != null) {
+                    tenantAwareUsername = user.getUsername();
+                    username = UserCoreUtil.addTenantDomainToEntry(tenantAwareUsername, user.getTenantDomain());
+                    userId = user.getUserID();
+                }
+            }
+        }
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:232]: can't we extract this out to a private method?

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 8]
```diff
@@ -149,6 +165,163 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+                    }
+                    throw new AuthenticationFailedException(
+                            MagicLinkAuthErrorConstants.ErrorMessages.INVALID_TENANT_ID_OF_THE_USER.getCode(),
+                            e.getMessage(),
+                            org.wso2.carbon.identity.application.common.model.User.getUserFromUserName(username), e);
+                } catch (org.wso2.carbon.user.api.UserStoreException e) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("IdentifierHandler failed while trying to authenticate", e);
+                    }
+                    throw new AuthenticationFailedException(
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:283]: can't we extract the username validation to a private method?

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/util/MagicLinkAuthErrorConstants.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+public class MagicLinkAuthErrorConstants {
+
+    /**
+     * Relevant error messages and error codes.
+     * This error codes are commented as a group not because of the convention
+     * because to maintain the togetherness of the errors.
+     */
+    public enum ErrorMessages {
+
+        // Identifier related Error codes
```
Comment (sahandilshan) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/util/MagicLinkAuthErrorConstants.java:12]: Missing the full stop. Fix in other places as well

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 10]
```diff
@@ -329,4 +403,322 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+                tenantAwareUsername, username, authProperties);
+        if (userIdFromOrganizationHierarchy.isPresent()) {
+            userId = userIdFromOrganizationHierarchy.get();
+        }
+        Optional<String> userIdFromUserStore = resolveUserFromUserStore(tenantDomain, tenantAwareUsername,
+                userId, username, authProperties, context);
+        if (userIdFromUserStore.isPresent()) {
+            userId = userIdFromUserStore.get();
+        }
+        if (userId == null) {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:466]:
```suggestion
        if (StringUtils.isNotBlank(userId)) {
```

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 11]
```diff
@@ -329,4 +403,322 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> resolveUserFromOrganizationHierarchy(AuthenticationContext context,
+                                                                  String tenantAwareUsername, String username,
+                                                                  Map<String, Object> authProperties)
+            throws AuthenticationFailedException {
+
+        if (!preconditionsForResolvingUserFromOrganizationHierarchy(context)) {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:569]: how about a method name that suggests it returns a boolean?

suggestion: canResolveUserFromOrganizationHierarchy

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticatorConstants.java [hunk 12]
```diff
@@ -35,9 +35,27 @@ private MagicLinkAuthenticatorConstants() {
...
+    public static final String VALIDATE_USERNAME = "ValidateUsername";
+
+    /**
+     * Define logging constants.
+     */
+    public static class LogConstants {
+
+        public static final String MAGICLINK_LOCAL_SERVICE = "magiclink-local-service";
+        public static final String FAILED = "FAILED";
+    }
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticatorConstants.java:57]: let's add these constants when we add diag logs

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/internal/MagicLinkServiceDataHolder.java [hunk 13]
```diff
@@ -61,6 +65,25 @@ public void setRealmService(RealmService realmService) {
...
+    }
+
+    public OrganizationUserResidentResolverService getOrganizationUserResidentResolverService() {
+
+        return organizationUserResidentResolverService;
+    }
+
+    public void setOrganizationUserResidentResolverService
+            (OrganizationUserResidentResolverService organizationUserResidentResolverService) {
+        this.organizationUserResidentResolverService = organizationUserResidentResolverService;
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/internal/MagicLinkServiceDataHolder.java:84]: checkstyle issue

Comment (chamathns) [reply components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/internal/MagicLinkServiceDataHolder.java:84]: also let's add method comments ^

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 14]
```diff
@@ -67,6 +88,28 @@ public class MagicLinkAuthenticator extends AbstractApplicationAuthenticator imp
...
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request,
+                                           HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
+        if (isIdentifierFirstRequest(request)) {
+            if (context.isLogoutRequest()) {
+                return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+            } else {
+                resolveUserFromIdfAuthenticationResponse(request, context);
+                if (getName().equals(context.getProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR))) {
```
[the remark this answers is not in the document]
Comment (Yoshani) [reply components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:100]: getName() is an overridden method

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 15]
```diff
@@ -329,4 +404,322 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+    /**
+     * Check if request type is identifier first.
+     *
+     * @param request HttpServletRequest.
+     * @return boolean.
+     */
+    private boolean isIdentifierFirstRequest(HttpServletRequest request) {
+
+        String authType = request.getParameter(AUTH_TYPE);
+        return IDF.equals(authType) || request.getParameter(IDENTIFIER_CONSENT) != null;
```
Comment (Yoshani) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:721]: check if identifier_consent is applicable

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 16]
```diff
@@ -329,4 +404,322 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+    }
+
+    /**
+     * This method is used to resolve the user from authentication response from identifier handler.
+     *
+     * @param request  The httpServletRequest.
+     * @param context  The authentication context.
+     * @throws AuthenticationFailedException In occasions of failing.
+     */
+    private void resolveUserFromIdfAuthenticationResponse(HttpServletRequest request, AuthenticationContext context)
```
Comment (Yoshani) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:420]: improve method name

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 17]
```diff
@@ -67,6 +82,30 @@ public class MagicLinkAuthenticator extends AbstractApplicationAuthenticator imp
     private static final long serialVersionUID = 4345354156955223654L;
     private static final Log log = LogFactory.getLog(MagicLinkAuthenticator.class);
 
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request,
+                                           HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException, LogoutFailedException {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:88]: Check for formatting here

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 18]
```diff
@@ -167,6 +234,15 @@ public String getFriendlyName() {
...
     public boolean canHandle(HttpServletRequest httpServletRequest) {
 
+        if (isIdentifierFirstRequest(httpServletRequest)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Magic link authenticator is handling identifier first flow ");
+            }
+            String userName = httpServletRequest.getParameter(MagicLinkAuthenticatorConstants.USER_NAME);
+            String restart = httpServletRequest.getParameter(RESTART_FLOW);
+
+            return userName != null || restart != null;
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:244]: let's check for empty as well

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 19]
```diff
@@ -259,41 +331,42 @@ private User getUser(AuthenticatedUser authenticatedUser) throws AuthenticationF
...
-                            log.debug("There are more than one user with the provided username claim value: "
-                                    + authenticatedUser.getUserName());
-                        }
-                        return null;
-                    }
-                    user = userList.get(0);
-                } else {
-                    log.error("Cannot find the user realm for the given tenant: " + tenantDomain);
+        if (tenantDomain == null) {
+            return null;
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:335]: can't we change the method to return an Optional here?

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 20]
```diff
@@ -329,4 +402,105 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+    private String getIdentifierFromRequest(HttpServletRequest request) {
+
+        return request.getParameter(MagicLinkAuthenticatorConstants.USER_NAME);
+    }
+
+    private void validateIdentifierFromRequest(HttpServletRequest request)
+            throws AuthenticationFailedException {
+
+        String identifierFromRequest = getIdentifierFromRequest(request);
+        if (StringUtils.isBlank(identifierFromRequest)) {
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:415]: We can use Boolean.parseBoolean here

Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:415]:
```suggestion

    private void validateIdentifierFromRequest(HttpServletRequest request)
            throws AuthenticationFailedException {

        if (Boolean.parseBoolean(
        request.getParameter(MagicLinkAuthenticatorConstants.USER_NAME)) {
```

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 21]
```diff
@@ -329,4 +402,105 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+        String tenantDomain = MultitenantUtils.getTenantDomain(username);
+
+        user.setUsername(tenantAwareUsername);
+        user.setUserStoreDomain(UserCoreUtil.extractDomainFromName(username));
+        user.setTenantDomain(tenantDomain);
+        return user;
+    }
+
+    private void setResolvedUserInContext(AuthenticationContext context, User user) {
+        Map<String, Object> authProperties = context.getProperties();
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:447]: Leave a blank line after method signature

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 22]
```diff
@@ -329,4 +402,90 @@ private List<String> getBlockedUserStoreDomainsList() {
...
+     *
+     * @param request  The httpServletRequest.
+     * @param context  The authentication context.
+     * @throws AuthenticationFailedException In occasions of failing.
+     */
+    private User resolveUser(HttpServletRequest request, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String username = validateIdentifierFromRequest(request);
+
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:428]: these blank lines seem redundant

**Review by chamathns — DISMISSED** [review]

**Review by chamathns — DISMISSED** [review]

**Comment by Yoshani** [comment]
## Manual Testing
Manual test round performed and verified.

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/test/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticatorTest.java [hunk 23]
```diff
@@ -396,17 +430,99 @@ public void testInitiateAuthenticationRequestWithIOException() throws Exception
...
+        Assert.assertEquals(status, AuthenticatorFlowStatus.INCOMPLETE);
+    }
+
     @ObjectFactory
     public IObjectFactory getObjectFactory() {
 
         return new PowerMockObjectFactory();
     }
-}
+}
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/test/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticatorTest.java:528]: Let's add a blank line

**Review by chamathns — APPROVED** [review]

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/pom.xml [hunk 24]
```diff
@@ -1,18 +1,20 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
-~ Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+~ Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/pom.xml:3]: Fix the licence header

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java [hunk 25]
```diff
@@ -66,6 +80,31 @@ public class MagicLinkAuthenticator extends AbstractApplicationAuthenticator imp
 
     private static final long serialVersionUID = 4345354156955223654L;
     private static final Log log = LogFactory.getLog(MagicLinkAuthenticator.class);
+    private AuthenticationContext authenticationContext;
+
+    @Override
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/MagicLinkAuthenticator.java:85]: Add method comment

File: pom.xml [hunk 26]
```diff
@@ -1,18 +1,20 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
-~ Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+~ Copyright (c) 2021, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (hwupathum) [inline pom.xml:3]: Fix the licence header

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/util/MagicLinkAuthErrorConstants.java [hunk 27]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/util/MagicLinkAuthErrorConstants.java:2]: Fix licence header

File: components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/internal/MagicLinkAuthenticatorServiceComponent.java [hunk 28]
```diff
@@ -1,7 +1,7 @@
 /*
- * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2021, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Yoshani) [reply components/org.wso2.carbon.identity.application.authenticator.magiclink/src/main/java/org/wso2/carbon/identity/application/authenticator/magiclink/internal/MagicLinkAuthenticatorServiceComponent.java:2]: Already fixed in latest version

**Comment by Yoshani** [comment]
Comments fixed by https://github.com/wso2-extensions/identity-local-auth-magiclink/pull/33

**Pull request merged**